src/js/utils/fullscreen-api.js

```diff
@@ -1,27 +1,179 @@
+import { WebkitPresentationModes } from '../constants.js';
+import { containsComposedNode } from './element-utils.js';
 import { document } from './server-safe-globals.js';
 
-export const fullscreenApi = {
-  enter: 'requestFullscreen',
-  exit: 'exitFullscreen',
-  rootEvents: ['fullscreenchange'],
-  mediaEvents: [],
-  element: 'fullscreenElement',
-  error: 'fullscreenerror',
-  enabled: 'fullscreenEnabled',
+// NOTE: (re)defining these types, but more narrowly for API expectations. These should probably be centralized + derived
+// once migrated to TypeScript types (CJP)
+
+/**
+ * @typedef {Partial<HTMLVideoElement> & {
+ *  webkitDisplayingFullscreen?: boolean;
+ *  webkitPresentationMode?: 'fullscreen'|'picture-in-picture';
+ *  webkitEnterFullscreen?: () => any;
+ * }} MediaStateOwner
+ */
+
+/**
+ * @typedef {Partial<Document|ShadowRoot>} RootNodeStateOwner
+ */
+
+/**
+ * @typedef {Partial<HTMLElement>} FullScreenElementStateOwner
+ */
+
+/**
+ * @typedef {object} StateOwners
+ * @property {MediaStateOwner} [media]
+ * @property {RootNodeStateOwner} [documentElement]
+ * @property {FullScreenElementStateOwner} [fullscreenElement]
+ */
+
+/** @type {(stateOwners: StateOwners) => Promise<undefined> | undefined} */
+export const enterFullscreen = (stateOwners) => {
```

Comment:
This is definitely a nit pick, so take it or leave it, but if we're keeping these as `Utils` it'd feel better if they weren't tied to the concept of stateOwners and media + fullscreenEl were specific arguments.

Reply:
Yeah I def get that intuition. Here was my reasoning:

- If you squint, this is just a common convention for passing in multiple optional arguments in JS (an object with optional key + value pairs).
- The "shape" of that argument object **_happens_** to use the same names as `StateOwners`, in large part out of convenience (since, aside from one case in platform checks, it's only used by the Media Store architecture)
- Even there, it isn't **_identical_** to `StateOwners`, because it only expects/requires a small subset of what has to be defined on them

I'm gonna leave for now but we can certainly revisit in next week's MCOH (feels like it would fit in the "best practices" category if we wanna define something either way here)